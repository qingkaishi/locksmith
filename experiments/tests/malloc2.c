#include <pthread.h>
#include <stdio.h>

pthread_t t1;

void * dostuff(void * a) {
  int j;
  int** b = (int**)a;
  (**b) = 1; // read on (2), write on (1)
  (*b) = NULL; // write on (2)
}

int id(int **a) { return **a;  } // read on (1),(2)

int main() {
  int *i = (int *) malloc(sizeof(int));    /* (1) */
  int ** a = (int**) malloc(sizeof(int*)); /* (2) */
  (*a) = i; // no read on (2); unique
  pthread_create(&t1, NULL, dostuff, a);
  id(a);
}


