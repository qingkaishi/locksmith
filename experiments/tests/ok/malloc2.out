************************* STARTING *************************

shared: alloc:malloc2.c:16
        alloc:malloc2.c:17

Warning: Possible data race: alloc:malloc2.c:16 is not protected!
 references:
  dereference of *b:malloc2.c:8 at malloc2.c:9
    alloc:malloc2.c:16 => tmp::-1
                       => i:malloc2.c:16
                       => *a:malloc2.c:17
                       => *cast:malloc2.c:8 at dostuff malloc2.c:19
                       => *b:malloc2.c:8
  locks acquired:
    <empty>
  in: main at malloc2.c:15 

  dereference of *a:malloc2.c:13 at malloc2.c:13
    alloc:malloc2.c:16 => tmp::-1
                       => i:malloc2.c:16
                       => *a:malloc2.c:17
                       => *id.a:malloc2.c:20
                       => *a:malloc2.c:13 at id malloc2.c:20
  locks acquired:
    <empty>
  in: main at malloc2.c:15  -> malloc2.c:20



Warning: Possible data race: alloc:malloc2.c:17 is not protected!
 references:
  dereference of b:malloc2.c:8 at malloc2.c:9
    alloc:malloc2.c:17 => tmp___0::-1
                       => a:malloc2.c:17
                       => cast:malloc2.c:8 at dostuff malloc2.c:19
                       => b:malloc2.c:8
  locks acquired:
    <empty>
  in: main at malloc2.c:15 

  dereference of b:malloc2.c:8 at malloc2.c:10
    alloc:malloc2.c:17 => tmp___0::-1
                       => a:malloc2.c:17
                       => cast:malloc2.c:8 at dostuff malloc2.c:19
                       => b:malloc2.c:8
  locks acquired:
    <empty>
  in: main at malloc2.c:15 

  dereference of a:malloc2.c:13 at malloc2.c:13
    alloc:malloc2.c:17 => tmp___0::-1
                       => a:malloc2.c:17
                       => id.a:malloc2.c:20
                       => a:malloc2.c:13 at id malloc2.c:20
  locks acquired:
    <empty>
  in: main at malloc2.c:15  -> malloc2.c:20


*************************** DONE ***************************
